dnf5/commands/offline/offline.cpp

```diff
@@ -0,0 +1,582 @@
+    }
+}
+
+void clean_datadir(Context & ctx, const std::filesystem::path & datadir) {
+    ctx.base.get_logger()->info("Cleaning up downloaded data...");
+
+    for (const auto & entry : std::filesystem::directory_iterator(datadir)) {
+        std::filesystem::remove_all(entry.path());
+    }
+}
+
+void OfflineRebootCommand::set_argument_parser() {
+    OfflineSubcommand::set_argument_parser();
+
+    auto & ctx = get_context();
+    auto & parser = ctx.get_argument_parser();
+    auto & cmd = *get_argument_parser_command();
+
+    cmd.set_description("Prepare the system to perform the offline transaction and reboots to start the transaction.");
+
+    poweroff_after =
+        dynamic_cast<libdnf5::OptionBool *>(parser.add_init_value(std::make_unique<libdnf5::OptionBool>(true)));
+
+    auto * poweroff_after_arg = parser.add_new_named_arg("poweroff");
+    poweroff_after_arg->set_long_name("poweroff");
+    poweroff_after_arg->set_description("Power off the system after the operation is complete");
+    poweroff_after_arg->link_value(poweroff_after);
+
+    cmd.register_named_arg(poweroff_after_arg);
+}
+
+void OfflineRebootCommand::run() {
+    auto & ctx = get_context();
+
+    check_state(*state);
+    if (state->get_data().status != libdnf5::offline::STATUS_DOWNLOAD_COMPLETE) {
+        throw libdnf5::cli::CommandExitError(1, M_("system is not ready for offline transaction"));
+    }
+    if (std::filesystem::is_symlink(get_magic_symlink())) {
+        throw libdnf5::cli::CommandExitError(1, M_("offline {} is already scheduled"), state->get_data().verb);
+    }
+
+    if (!std::filesystem::is_directory(get_datadir())) {
+        throw libdnf5::cli::CommandExitError(1, M_("data directory {} does not exist"), get_datadir().string());
+    }
+
+    if (state->get_data().verb == "system-upgrade download") {
+        std::cout << "The system will now reboot to upgrade to release version " << state->get_data().target_releasever
+                  << "." << std::endl;
+    } else {
+        std::cout << "The system will now reboot to perform the offline transaction initiated by the following command:"
+                  << std::endl
+                  << "\t" << state->get_data().cmd_line << std::endl;
+    }
+    if (!libdnf5::cli::utils::userconfirm::userconfirm(ctx.base.get_config())) {
+        return;
+    }
+
+    std::filesystem::create_symlink(get_datadir(), get_magic_symlink());
+
+    state->get_data().status = libdnf5::offline::STATUS_READY;
+    state->get_data().poweroff_after = poweroff_after->get_value();
+    state->write();
+
+    reboot(poweroff_after->get_value());
+}
+
+void OfflineExecuteCommand::set_argument_parser() {
+    OfflineSubcommand::set_argument_parser();
+    auto & cmd = *get_argument_parser_command();
+    cmd.set_description(
+        "Internal use only, not intended to be run by the user. Executes the transaction in the offline environment.");
```

Comment:
I set `set_complete(false)`. The `visible` flag could be nice, but I don't think it's a problem that it's documented in the help, it could help someone debug `dnf5-offline-transaction.service`. I think it's clear enough that it's not meant to be used.